OpenRA.Mods.AS/Traits/Nydus.cs

```diff
@@ -0,0 +1,207 @@
+			if (counter.Count == 0)
+			{
+				var pri = self.Trait<NydusPrimaryExit>();
+				pri.SetPrimary(self);
+			}
+
+			counter.Count++;
+		}
+
+		void DecreaseNydusCnt(Actor self, Player owner)
+		{
+			var counter = owner.PlayerActor.Trait<NydusCounter>();
+			counter.Count--;
+
+			// what if primary was killed?
+			if (self.IsPrimaryNydusExit())
+			{
+				var actors = self.World.ActorsWithTrait<NydusPrimaryExit>()
+					.Where(a => a.Actor.Owner == self.Owner && a.Actor != self);
+				if (!actors.Any())
+				{
+					// no nydus canal left.
+					counter.PrimaryActor = null;
+				}
+				else
+				{
+					var pri = actors.First().Actor;
+					pri.Trait<NydusPrimaryExit>().SetPrimary(pri);
+				}
+			}
+		}
+
+		public void AddedToWorld(Actor self) { IncreaseNydusCnt(self, self.Owner); }
+		public void Disposing(Actor self) { DecreaseNydusCnt(self, self.Owner); }
+
+		public void OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
+		{
+			// this one is trickier haha.
+			IncreaseNydusCnt(self, newOwner);
+			DecreaseNydusCnt(self, oldOwner);
+		}
+	}
+
+	[Desc("Can move instantly to primary designated nydus canal actors.")]
+	class NydusTransportableInfo : ITraitInfo
+	{
+		[VoiceReference]
+		public readonly string Voice = "Action";
+		public readonly string EnterCursor = "enter";
+		public readonly string EnterBlockedCursor = "enter-blocked";
+		public object Create(ActorInitializer init) { return new NydusTransportable(init, this); }
+	}
+
+	class NydusTransportable : IIssueOrder, IResolveOrder, IOrderVoice
+	{
+		readonly NydusTransportableInfo info;
+
+		public NydusTransportable(ActorInitializer init, NydusTransportableInfo info)
+		{
+			this.info = info;
+		}
+
+		public IEnumerable<IOrderTargeter> Orders
+		{
+			get { yield return new NydusTransportOrderTargeter(info); }
+		}
+
+		public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool queued)
+		{
+			if (order.OrderID != "NydusTransport")
+				return null;
+
+			// You can't enter enemy nydus canal so this will do.
+			if (target.Type == TargetType.FrozenActor)
+				return null;
+
+			return new Order(order.OrderID, self, target, queued) { };
+		}
+
+		// Does the player own enough nydus canals?
+		// For nydus canals to work, the player needs at least 2.
+		// Actually, at least 1 but in that case, the unit goes in and immediately comes out. That's stupid.
+		static bool HasEnoughCanals(Actor self)
+		{
+			var counter = self.Owner.PlayerActor.Trait<NydusCounter>();
+			return counter.Count > 1;
+		}
+
+		static bool IsValidOrder(Actor self, Order order)
+		{
+			// Not targeting a frozen actor
+			if (order.ExtraData == 0 && order.TargetActor == null)
+				return false;
+
+			// The owner must own at least two nydus canals.
+			if (!HasEnoughCanals(self))
+				return false;
+
+			// primary buildings are where entered units exit.
+			// You don't want to put them in the primary building.
+			return !order.TargetActor.IsPrimaryNydusExit();
```

Comment:
Notsure on this one - could very well be that a palyer is timing his primary building setting to onyl change it about the time the actor would reach the nydus.